README
Hi all, this is my first attempt at qt programming, HypeRoam is still in its pre-alpha (:P) stage so you may encounter lot of bugs...
and yeah put on your crash helmet before trying it :)
mail me ravirdv [at] gmail [dot] com for any kind of feedback,  feature request or suggestions ;)


PS: you do not need to add "da-iict.org" when adding accounts :)

have fun!


Changelog
0.2
	* Settings are saved using QSettings
	* New Icons :D
	* Many bugfixes
	* Middle click to change account
	* Added Tray Icon 
0.1 
	* Basic Functionality in place :D 
	* Initial  UI Design


main.cpp
/*
    HyperRoam - CybeRoam client with automatic account switching
    Copyright (C) 2009-10  Ravi Vagadia

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
//Qt
#include <QtGui/QApplication>
#include <QIcon>
#include <QPixmap>
//HypeRoam
#include "widget.h"

int main(int argc, char *argv[])
{
    QApplication a(argc, argv);
    QIcon icon(":/icons/app");
    Widget w;
    w.setWindowTitle("HypeRoam");
    w.show();
    a.setApplicationName("HypeRoam");
    a.setWindowIcon(icon);
    a.setApplicationVersion("0.2 PreBeta");
    a.setOrganizationName("Invalid Technologies");
    a.setQuitOnLastWindowClosed(false);
    return a.exec();
}


widget.h
#ifndef WIDGET_H
#define WIDGET_H

#include <QWidget>
#include <QNetworkAccessManager>
#include <QNetworkRequest>
#include <QUrl>
#include <QNetworkReply>
#include <QSqlDatabase>
#include <QSystemTrayIcon>
#include <database.h>
#include <QListWidgetItem>
#include <QTimer>
#include <QMenu>
#include <QAction>
#include <QSettings>

namespace Ui {
    class Widget;
}

class Widget : public QWidget {
    Q_OBJECT
public:
    int Opacity, idIndex, failCounter;
    unsigned int sleepTime;
    Widget(QWidget *parent = 0);
    ~Widget();
    QNetworkAccessManager *manager,;
    QNetworkRequest req;
    QNetworkReply *reply, ;
    QString uid,pwd;
    QSystemTrayIcon *mTray;
    database db;
    QTimer *timer, *timer1;
    QMenu *trayIconMenu;
    QAction *minimizeAction, *restoreAction, *quitAction,*switchAction;
    QSettings *settings;
    bool disableBalloonTips;
protected:
    void changeEvent(QEvent *e);

private:
    Ui::Widget *ui;

    void fillList();
    void setupTray();
    void Notify(QString str);
    void initOpacity();

private slots:
    void on_autoSwitch_toggled(bool checked);
    void on_switchTimeSlider_valueChanged(int value);
    void on_disableBalloonTips_toggled(bool checked);
    void tryConnect(QString id, QString pwd);
    void on_checkBox_toggled(bool checked);
    void on_accountList_itemDoubleClicked(QListWidgetItem* item);
    void on_switchAccount_clicked();
    void on_removeAccount_clicked();
    void on_startButton_clicked();
    void on_addButton_clicked();
    void loginFinished(QNetworkReply*);
    void changeOpacity();
    void trayActivated(QSystemTrayIcon::ActivationReason reason);
};

#endif // WIDGET_H


ui_widget.h
/********************************************************************************
** Form generated from reading UI file 'widget.ui'
**
** Created: Wed Mar 31 19:45:19 2010
**      by: Qt User Interface Compiler version 4.6.2
**
** WARNING! All changes made in this file will be lost when recompiling UI file!
********************************************************************************/

#ifndef UI_WIDGET_H
#define UI_WIDGET_H

#include <QtCore/QVariant>
#include <QtGui/QAction>
#include <QtGui/QApplication>
#include <QtGui/QButtonGroup>
#include <QtGui/QCheckBox>
#include <QtGui/QFrame>
#include <QtGui/QHeaderView>
#include <QtGui/QLabel>
#include <QtGui/QLineEdit>
#include <QtGui/QListWidget>
#include <QtGui/QPushButton>
#include <QtGui/QSlider>
#include <QtGui/QWidget>

QT_BEGIN_NAMESPACE

class Ui_Widget
{
public:
    QPushButton *addButton;
    QLabel *label;
    QLineEdit *username;
    QLabel *label_2;
    QLineEdit *password;
    QPushButton *startButton;
    QListWidget *accountList;
    QLabel *label_3;
    QLabel *label_4;
    QLabel *inUseLabel;
    QPushButton *removeAccount;
    QPushButton *switchAccount;
    QLabel *label_5;
    QLabel *notificationLabel;
    QFrame *line;
    QFrame *line_2;
    QFrame *line_3;
    QLabel *label_8;
    QCheckBox *checkBox;
    QCheckBox *autoSwitch;
    QCheckBox *disableBalloonTips;
    QSlider *switchTimeSlider;
    QLabel *label_6;

    void setupUi(QWidget *Widget)
    {
        if (Widget->objectName().isEmpty())
            Widget->setObjectName(QString::fromUtf8("Widget"));
        Widget->resize(405, 343);
        QSizePolicy sizePolicy(QSizePolicy::Fixed, QSizePolicy::Fixed);
        sizePolicy.setHorizontalStretch(0);
        sizePolicy.setVerticalStretch(0);
        sizePolicy.setHeightForWidth(Widget->sizePolicy().hasHeightForWidth());
        Widget->setSizePolicy(sizePolicy);
        Widget->setMinimumSize(QSize(405, 343));
        Widget->setMaximumSize(QSize(405, 343));
        addButton = new QPushButton(Widget);
        addButton->setObjectName(QString::fromUtf8("addButton"));
        addButton->setGeometry(QRect(90, 99, 95, 24));
        label = new QLabel(Widget);
        label->setObjectName(QString::fromUtf8("label"));
        label->setGeometry(QRect(21, 54, 61, 16));
        username = new QLineEdit(Widget);
        username->setObjectName(QString::fromUtf8("username"));
        username->setGeometry(QRect(88, 50, 121, 25));
        username->setMaxLength(32);
        label_2 = new QLabel(Widget);
        label_2->setObjectName(QString::fromUtf8("label_2"));
        label_2->setGeometry(QRect(21, 77, 61, 16));
        password = new QLineEdit(Widget);
        password->setObjectName(QString::fromUtf8("password"));
        password->setGeometry(QRect(88, 73, 121, 25));
        password->setMaxLength(64);
        password->setEchoMode(QLineEdit::Password);
        startButton = new QPushButton(Widget);
        startButton->setObjectName(QString::fromUtf8("startButton"));
        startButton->setGeometry(QRect(10, 230, 95, 24));
        accountList = new QListWidget(Widget);
        accountList->setObjectName(QString::fromUtf8("accountList"));
        accountList->setGeometry(QRect(230, 50, 161, 141));
        label_3 = new QLabel(Widget);
        label_3->setObjectName(QString::fromUtf8("label_3"));
        label_3->setGeometry(QRect(230, 197, 61, 16));
        label_4 = new QLabel(Widget);
        label_4->setObjectName(QString::fromUtf8("label_4"));
        label_4->setGeometry(QRect(230, 30, 61, 16));
        inUseLabel = new QLabel(Widget);
        inUseLabel->setObjectName(QString::fromUtf8("inUseLabel"));
        inUseLabel->setGeometry(QRect(300, 197, 101, 16));
        removeAccount = new QPushButton(Widget);
        removeAccount->setObjectName(QString::fromUtf8("removeAccount"));
        removeAccount->setGeometry(QRect(90, 126, 95, 24));
        switchAccount = new QPushButton(Widget);
        switchAccount->setObjectName(QString::fromUtf8("switchAccount"));
        switchAccount->setGeometry(QRect(20, 20, 101, 24));
        label_5 = new QLabel(Widget);
        label_5->setObjectName(QString::fromUtf8("label_5"));
        label_5->setGeometry(QRect(20, 164, 111, 16));
        notificationLabel = new QLabel(Widget);
        notificationLabel->setObjectName(QString::fromUtf8("notificationLabel"));
        notificationLabel->setGeometry(QRect(20, 280, 371, 20));
        notificationLabel->setStyleSheet(QString::fromUtf8("background-color: rgba(255, 85, 127, 0);"));
        notificationLabel->setAlignment(Qt::AlignCenter);
        line = new QFrame(Widget);
        line->setObjectName(QString::fromUtf8("line"));
        line->setGeometry(QRect(0, 268, 401, 16));
        line->setFrameShape(QFrame::HLine);
        line->setFrameShadow(QFrame::Sunken);
        line_2 = new QFrame(Widget);
        line_2->setObjectName(QString::fromUtf8("line_2"));
        line_2->setGeometry(QRect(0, 1, 401, 16));
        line_2->setFrameShape(QFrame::HLine);
        line_2->setFrameShadow(QFrame::Sunken);
        line_3 = new QFrame(Widget);
        line_3->setObjectName(QString::fromUtf8("line_3"));
        line_3->setGeometry(QRect(0, 300, 401, 16));
        line_3->setFrameShape(QFrame::HLine);
        line_3->setFrameShadow(QFrame::Sunken);
        label_8 = new QLabel(Widget);
        label_8->setObjectName(QString::fromUtf8("label_8"));
        label_8->setEnabled(false);
        label_8->setGeometry(QRect(277, 318, 121, 20));
        checkBox = new QCheckBox(Widget);
        checkBox->setObjectName(QString::fromUtf8("checkBox"));
        checkBox->setGeometry(QRect(246, 246, 131, 23));
        autoSwitch = new QCheckBox(Widget);
        autoSwitch->setObjectName(QString::fromUtf8("autoSwitch"));
        autoSwitch->setGeometry(QRect(247, 225, 131, 23));
        disableBalloonTips = new QCheckBox(Widget);
        disableBalloonTips->setObjectName(QString::fromUtf8("disableBalloonTips"));
        disableBalloonTips->setGeometry(QRect(10, 310, 141, 23));
        switchTimeSlider = new QSlider(Widget);
        switchTimeSlider->setObjectName(QString::fromUtf8("switchTimeSlider"));
        switchTimeSlider->setGeometry(QRect(20, 193, 160, 20));
        switchTimeSlider->setMinimum(10);
        switchTimeSlider->setMaximum(90);
        switchTimeSlider->setOrientation(Qt::Horizontal);
        label_6 = new QLabel(Widget);
        label_6->setObjectName(QString::fromUtf8("label_6"));
        label_6->setGeometry(QRect(130, 164, 91, 16));
        QWidget::setTabOrder(username, password);
        QWidget::setTabOrder(password, addButton);
        QWidget::setTabOrder(addButton, removeAccount);
        QWidget::setTabOrder(removeAccount, startButton);
        QWidget::setTabOrder(startButton, checkBox);
        QWidget::setTabOrder(checkBox, switchAccount);
        QWidget::setTabOrder(switchAccount, accountList);

        retranslateUi(Widget);

        QMetaObject::connectSlotsByName(Widget);
    } // setupUi

    void retranslateUi(QWidget *Widget)
    {
        Widget->setWindowTitle(QApplication::translate("Widget", "Widget", 0, QApplication::UnicodeUTF8));
        addButton->setText(QApplication::translate("Widget", "Add ->", 0, QApplication::UnicodeUTF8));
        label->setText(QApplication::translate("Widget", "Username", 0, QApplication::UnicodeUTF8));
        label_2->setText(QApplication::translate("Widget", "Password", 0, QApplication::UnicodeUTF8));
        password->setInputMask(QString());
        startButton->setText(QApplication::translate("Widget", "Login", 0, QApplication::UnicodeUTF8));
        label_3->setText(QApplication::translate("Widget", "In Use :", 0, QApplication::UnicodeUTF8));
        label_4->setText(QApplication::translate("Widget", "Accounts:", 0, QApplication::UnicodeUTF8));
        inUseLabel->setText(QApplication::translate("Widget", "None", 0, QApplication::UnicodeUTF8));
        removeAccount->setText(QApplication::translate("Widget", "Remove", 0, QApplication::UnicodeUTF8));
        switchAccount->setText(QApplication::translate("Widget", "Switch Account", 0, QApplication::UnicodeUTF8));
        label_5->setText(QApplication::translate("Widget", "Auto Switch Time", 0, QApplication::UnicodeUTF8));
        notificationLabel->setText(QApplication::translate("Widget", "Notifications", 0, QApplication::UnicodeUTF8));
        label_8->setText(QApplication::translate("Widget", "Author: HyperDUDe", 0, QApplication::UnicodeUTF8));
        checkBox->setText(QApplication::translate("Widget", "Connect on Starup", 0, QApplication::UnicodeUTF8));
        autoSwitch->setText(QApplication::translate("Widget", "Auto Switch", 0, QApplication::UnicodeUTF8));
        disableBalloonTips->setText(QApplication::translate("Widget", "Disable balloon tips", 0, QApplication::UnicodeUTF8));
        label_6->setText(QApplication::translate("Widget", "Minutes", 0, QApplication::UnicodeUTF8));
    } // retranslateUi

};

namespace Ui {
    class Widget: public Ui_Widget {};
} // namespace Ui

QT_END_NAMESPACE

#endif // UI_WIDGET_H


database.cpp
/*
    HyperRoam - CybeRoam client with automatic account switching
    Copyright (C) 2009-10  Ravi Vagadia

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

//HypeRoam
#include <database.h>
//Qt
#include <QDebug>
#include <QSqlError>
#include <QMessageBox>

/*Database Code is not removed just in case someone wants to use RDBMS to store login credentials */

database::database()
{
    settings = new QSettings("HyperTech", "HypeRoam");
  /*  if (db.isDriverAvailable("QSQLITE"))
    {
    db = QSqlDatabase::addDatabase("QSQLITE");
    db.setDatabaseName("data.db");
    db.open();
    db.isDriverAvailable("QSQLITE");
    initDatabase();

    }
    else
    {
       QMessageBox msg;
       msg.setText("QSLITE Driver Not Available");
       msg.exec();
    }
    */
 }

void database::addUser(QString uid, QString pwd)
{
      /*      QSqlQuery query;
            query.prepare("INSERT INTO users VALUES (:uid, :password)");
            query.bindValue(":uid", uid);
            query.bindValue(":password", pwd);
            if(query.exec())  qDebug()<<"User Added "<< uid;
      */
       settings->setValue(QString("users/").append(uid),pwd);
}


QString database::getPassword(QString uid)
{
    /*QSqlQuery query;
    query.prepare("select password from users where uname = :uid");
    query.bindValue(":uid", uid);
    query.exec();
    query.next();
    return query.value(0).toString();
    */
    return settings->value(QString("users/").append(uid)).toString();
}

void database::removeUser(QString uid)
{
            /*QSqlQuery query;
            query.prepare("delete from users where uname =:uid");
            query.bindValue(":uid", uid);
            query.exec();
            qDebug()<< "User Nuked :" << uid;
            */
    settings->remove(QString("users/").append(uid));
}

QSqlQuery database::populate()
{

    QSqlQuery query("SELECT * FROM users");
    return query;
}

int database::getRandom(int max)
{
    //srand ( time(NULL) );
    return rand() % max + 1;
}

void database::initDatabase()
{
    QSqlQuery query("create table if not exists users (uname text, password text);");
    query.exec();
}

bool database::userExists(QString uid)
{
    bool exists = false;
    settings->sync();
    settings->beginGroup("users");
    QStringList keys = settings->allKeys();
    foreach (QString str, keys)
      {
        if (uid.compare(str) == 0)
        {
           exists = true;
        }

       }

    settings->endGroup();
    qDebug()<<exists;
    return exists;
}


moc_widget.cpp
/****************************************************************************
** Meta object code from reading C++ file 'widget.h'
**
** Created: Mon Mar 29 02:01:35 2010
**      by: The Qt Meta Object Compiler version 62 (Qt 4.6.2)
**
** WARNING! All changes made in this file will be lost!
*****************************************************************************/

#include "widget.h"
#if !defined(Q_MOC_OUTPUT_REVISION)
#error "The header file 'widget.h' doesn't include <QObject>."
#elif Q_MOC_OUTPUT_REVISION != 62
#error "This file was generated using the moc from 4.6.2. It"
#error "cannot be used with the include files from this version of Qt."
#error "(The moc has changed too much.)"
#endif

QT_BEGIN_MOC_NAMESPACE
static const uint qt_meta_data_Widget[] = {

 // content:
       4,       // revision
       0,       // classname
       0,    0, // classinfo
      16,   14, // methods
       0,    0, // properties
       0,    0, // enums/sets
       0,    0, // constructors
       0,       // flags
       0,       // signalCount

 // slots: signature, parameters, type, tag, flags
      16,    8,    7,    7, 0x08,
      50,   44,    7,    7, 0x08,
      88,    8,    7,    7, 0x08,
     127,  124,    7,    7, 0x08,
     153,  146,    7,    7, 0x08,
     181,    7,    7,    7, 0x08,
     201,    8,    7,    7, 0x08,
     232,  227,    7,    7, 0x08,
     283,    7,    7,    7, 0x08,
     310,    7,    7,    7, 0x08,
     337,    7,    7,    7, 0x08,
     362,    7,    7,    7, 0x08,
     385,    7,    7,    7, 0x08,
     415,    7,    7,    7, 0x08,
     446,    7,    7,    7, 0x08,
     469,  462,    7,    7, 0x08,

       0        // eod
};

static const char qt_meta_stringdata_Widget[] = {
    "Widget\0\0checked\0on_autoSwitch_toggled(bool)\0"
    "value\0on_switchTimeSlider_valueChanged(int)\0"
    "on_disableBalloonTips_toggled(bool)\0"
    "id\0tryLogout(QString)\0id,pwd\0"
    "tryConnect(QString,QString)\0"
    "on_logout_clicked()\0on_checkBox_toggled(bool)\0"
    "item\0on_accountList_itemDoubleClicked(QListWidgetItem*)\0"
    "on_switchAccount_clicked()\0"
    "on_removeAccount_clicked()\0"
    "on_startButton_clicked()\0"
    "on_addButton_clicked()\0"
    "loginFinished(QNetworkReply*)\0"
    "logoutFinished(QNetworkReply*)\0"
    "changeOpacity()\0reason\0"
    "trayActivated(QSystemTrayIcon::ActivationReason)\0"
};

const QMetaObject Widget::staticMetaObject = {
    { &QWidget::staticMetaObject, qt_meta_stringdata_Widget,
      qt_meta_data_Widget, 0 }
};

#ifdef Q_NO_DATA_RELOCATION
const QMetaObject &Widget::getStaticMetaObject() { return staticMetaObject; }
#endif //Q_NO_DATA_RELOCATION

const QMetaObject *Widget::metaObject() const
{
    return QObject::d_ptr->metaObject ? QObject::d_ptr->metaObject : &staticMetaObject;
}

void *Widget::qt_metacast(const char *_clname)
{
    if (!_clname) return 0;
    if (!strcmp(_clname, qt_meta_stringdata_Widget))
        return static_cast<void*>(const_cast< Widget*>(this));
    return QWidget::qt_metacast(_clname);
}

int Widget::qt_metacall(QMetaObject::Call _c, int _id, void **_a)
{
    _id = QWidget::qt_metacall(_c, _id, _a);
    if (_id < 0)
        return _id;
    if (_c == QMetaObject::InvokeMetaMethod) {
        switch (_id) {
        case 0: on_autoSwitch_toggled((*reinterpret_cast< bool(*)>(_a[1]))); break;
        case 1: on_switchTimeSlider_valueChanged((*reinterpret_cast< int(*)>(_a[1]))); break;
        case 2: on_disableBalloonTips_toggled((*reinterpret_cast< bool(*)>(_a[1]))); break;
        case 3: tryLogout((*reinterpret_cast< QString(*)>(_a[1]))); break;
        case 4: tryConnect((*reinterpret_cast< QString(*)>(_a[1])),(*reinterpret_cast< QString(*)>(_a[2]))); break;
        case 5: on_logout_clicked(); break;
        case 6: on_checkBox_toggled((*reinterpret_cast< bool(*)>(_a[1]))); break;
        case 7: on_accountList_itemDoubleClicked((*reinterpret_cast< QListWidgetItem*(*)>(_a[1]))); break;
        case 8: on_switchAccount_clicked(); break;
        case 9: on_removeAccount_clicked(); break;
        case 10: on_startButton_clicked(); break;
        case 11: on_addButton_clicked(); break;
        case 12: loginFinished((*reinterpret_cast< QNetworkReply*(*)>(_a[1]))); break;
        case 13: logoutFinished((*reinterpret_cast< QNetworkReply*(*)>(_a[1]))); break;
        case 14: changeOpacity(); break;
        case 15: trayActivated((*reinterpret_cast< QSystemTrayIcon::ActivationReason(*)>(_a[1]))); break;
        default: ;
        }
        _id -= 16;
    }
    return _id;
}
QT_END_MOC_NAMESPACE


widget.cpp
/*
    HyperRoam - CybeRoam client with automatic account switching
    Copyright (C) 2009-10  Ravi Vagadia

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

//HypeRoam

#include "widget.h"
#include "ui_widget.h"

//Qt
#include <QDebug>
#include <QByteArray>
#include <QSqlQuery>
#include <QThread>


Widget::Widget(QWidget *parent) :
    QWidget(parent),
    ui(new Ui::Widget)
{
    ui->setupUi(this);
    manager = new QNetworkAccessManager();
    timer = new QTimer(this);
    timer1 = new QTimer(this);
    connect(manager,SIGNAL(finished(QNetworkReply*)),this, SLOT(loginFinished(QNetworkReply*)));
    connect(timer, SIGNAL(timeout()),this, SLOT(on_switchAccount_clicked()));
    connect(timer1, SIGNAL(timeout()),this, SLOT(changeOpacity()));

    setupTray();
    settings = new QSettings("HyperTech", "HypeRoam");
    req.setUrl(QUrl("http://10.100.56.55:8090/corporate/servlet/CyberoamHTTPClient"));


    //ui->switchTimeSlider->setValue(settings->value("Time").toInt());


    //Stuff related to QSettings
        ui->disableBalloonTips->setChecked(settings->value("disableBalloonTips").toBool());
        ui->autoSwitch->setChecked(settings->value("autoSwitch").toBool());
        if(settings->value("AutoConnect").toBool() && ui->accountList->count())
            {
                ui->checkBox->setChecked(true);
                on_switchAccount_clicked();
                QTimer::singleShot(0, this, SLOT(hide()));
            }
//Initializations
    idIndex = 0;
    failCounter = 0;
    sleepTime = 3000;
    initOpacity();
    fillList();
    ui->switchAccount->setVisible(false);
    ui->label_6->setText(QString::number(ui->switchTimeSlider->value()));
}

Widget::~Widget()
{
    delete ui;
}

void Widget::changeEvent(QEvent *e)
{
    QWidget::changeEvent(e);
    switch (e->type()) {
    case QEvent::LanguageChange:
        ui->retranslateUi(this);
        break;
    case QEvent::WindowStateChange:
         if( isMinimized() )
             QTimer::singleShot(0, this, SLOT(hide()));
        break;
    case QEvent::Quit:
            qDebug()<<"tesT";
            //emit aboutToQuit();
        break;
    default:
        break;
    }
}

void Widget::on_addButton_clicked()
{
   if(!db.userExists(ui->username->text()))
    {
        db.addUser(ui->username->text(),ui->password->text());
        fillList();
    }else
   {
        Notify("User Already in List");
    }
}

void Widget::on_startButton_clicked()
{
        if(ui->accountList->count() > 0)
        {

            if(ui->accountList->currentItem())
            {
                 uid = ui->accountList->currentItem()->text();
                 pwd = db.getPassword(uid);
                 tryConnect(uid,pwd);
                 if(ui->autoSwitch->isChecked() && !timer->isActive())
                  {
                      timer->start(ui->switchTimeSlider->value() * 60000);
                      hide();
                   }
            }
            else
            {
                 uid = ui->accountList->itemAt(0,0)->text();
                 pwd = db.getPassword(uid);
                 tryConnect(uid,pwd);
                     if(ui->autoSwitch->isChecked() && !timer->isActive())
                      {
                          timer->start(ui->switchTimeSlider->value() * 60000);
                          hide();
                       }
             }
        }
        else
        {
            Notify("Please Add an account");
        }

}

void Widget::loginFinished(QNetworkReply*)
{
        QString s = reply->readAll();
        //qDebug()<< s;
        if (s.contains("have"))
        {
            Notify(QString("You Have Successfully Logged in using ").append(uid));
            ui->inUseLabel->setText(uid);
             mTray->setIcon ( QIcon(":/icons/connected") );
            mTray->setToolTip(QString("HypeRoam - Connected using ").append(uid));
            failCounter = 0;
            sleepTime = 3000;
        }
        else if(s.contains("Multiple+login+not+allowed"))
        {
            ui->inUseLabel->setText("Multiple login not allowed");
            failCounter++;
            if(failCounter < 10) QTimer::singleShot(4000, this, SLOT(on_switchAccount_clicked()));
        }
        else
        {
            QTimer::singleShot(3000,this,SLOT());

            Notify("Some error :S");
            qDebug()<<s;
            mTray->setIcon( QIcon(":/icons/switch"));
            if (failCounter < 10 )
            {
                    QTimer::singleShot(sleepTime, this, SLOT(on_switchAccount_clicked()));
                    sleepTime = sleepTime * 2;
                    Notify("Account Automatically Switched");
                    failCounter++;
             }

        }
 }


void Widget::tryConnect(QString id, QString pwd)
{
    reply = manager->post(req, (QString("mode=191&isAccessDenied=null&url=null&message=&username=").append(id).append("@da-iict.org&password=").append(pwd).append("&saveinfo=saveinfo&login=Login")).toLocal8Bit());
}


void Widget::fillList()
{
    ui->accountList->clear();
    settings->sync();
    settings->beginGroup("users");
    QStringList keys = settings->allKeys();
    foreach (QString str, keys)
      {
                      ui->accountList->addItem(str);
       }
    settings->endGroup();
}

void Widget::on_removeAccount_clicked()
{
            db.removeUser(ui->accountList->currentItem()->text());
            fillList();
}

void Widget::on_switchAccount_clicked()
{
    if(ui->accountList->count())
    {
    if (idIndex == ui->accountList->count() - 1)
            idIndex = 0;
    else
            idIndex++;
    uid = ui->accountList->item(idIndex)->text();
    tryConnect(uid,db.getPassword(uid));
    if(ui->autoSwitch->isChecked())
    {
    timer->start(ui->switchTimeSlider->value() * 60000);
    }
    Notify(QString("Trying to Connect using ").append(uid));
    }
}

void Widget::setupTray()
{
    minimizeAction = new QAction(tr("Mi&nimize"), this);
    restoreAction = new QAction(tr("&Restore"), this);
    quitAction = new QAction(tr("&Quit"), this);
    switchAction = new QAction(tr("&Switch Account"), this);
    connect(minimizeAction, SIGNAL(triggered()), this, SLOT(hide()));
    connect(restoreAction, SIGNAL(triggered()), this, SLOT(showNormal()));
    connect(switchAction, SIGNAL(triggered()), this, SLOT(on_switchAccount_clicked()));
    connect(quitAction, SIGNAL(triggered()), qApp, SLOT(quit()));
        mTray = new QSystemTrayIcon( this );
        connect(mTray, SIGNAL(activated(QSystemTrayIcon::ActivationReason)),
                  this, SLOT(trayActivated(QSystemTrayIcon::ActivationReason)));

        mTray->setIcon ( QIcon(":/icons/app") );
        mTray->setToolTip("HypeRoam");
        trayIconMenu = new QMenu(this);
        trayIconMenu->addAction(minimizeAction);
        trayIconMenu->addAction(restoreAction);
        trayIconMenu->addSeparator();
        trayIconMenu->addAction(switchAction);
        trayIconMenu->addAction(quitAction);
        mTray->setContextMenu(trayIconMenu);
        mTray->show();
}

void Widget::on_accountList_itemDoubleClicked(QListWidgetItem* item)
{
    uid =item->text();
    pwd =db.getPassword(item->text());
    tryConnect(uid,pwd);
}

void Widget::changeOpacity()
{
   if((Opacity > 0))
    {
        ui->notificationLabel->setStyleSheet(QString("background-color: rgba(255, 85, 127, ").append(QString::number(Opacity -=2).append("%);")));
    }
    else
    {
        timer1->stop();
    }
}

void Widget::Notify(QString str)
{
    ui->notificationLabel->setText(str);
    ui->notificationLabel->setStyleSheet(QString("background-color: rgba(255, 85, 127, 100%);"));
    if(!Widget::isVisible() && !ui->disableBalloonTips->checkState())mTray->showMessage(QString("HypeRoam"), str, QSystemTrayIcon::Information , 2000);
    initOpacity();
    timer1->start(100);
}

void Widget::initOpacity()
{
    Opacity = 100;
}

void Widget::on_checkBox_toggled(bool checked)
{
    if(checked)
    {
        Notify("Auto Connect Mode Enabled");
        settings->setValue("AutoConnect",true);
    }
    else
    {
        Notify("Auto Connect Mode Disabled");
        settings->setValue("AutoConnect",false);
    }
}


void Widget::trayActivated(QSystemTrayIcon::ActivationReason reason)
{
    switch (reason) {
        case QSystemTrayIcon::DoubleClick:
             restoreAction->trigger();
             break;
         case QSystemTrayIcon::MiddleClick:
             on_switchAccount_clicked();
             break;

}
}

void Widget::on_disableBalloonTips_toggled(bool checked)
{
    if(checked)
    {
        settings->setValue("disableBalloonTips",true);
    }
    else
    {
        settings->setValue("disableBalloonTips",false);
    }
}



void Widget::on_switchTimeSlider_valueChanged(int value)
{
    ui->label_6->setText(QString::number(value));
    settings->setValue("Time",ui->switchTimeSlider->value());
}



void Widget::on_autoSwitch_toggled(bool checked)
{
    if(checked)
    {
        settings->setValue("autoSwitch",true);
    }
    else
    {
        settings->setValue("autoSwitch",false);
    }
}
